# Installs
### Uncomment to install dependencies

In [1]:
# !python.exe -m pip install --upgrade pip
# !pip install pandas
# !pip install numpy
# !pip install altair
# !pip install "vegafusion[embed]"   
# !pip install vl-convert-python>=1.6.0
# !pip install pyarrow

# Imports

In [2]:
import pandas as pd
import numpy as np
import altair as alt
import vegafusion as vf

alt.data_transformers.enable("vegafusion")

DataTransformerRegistry.enable('vegafusion')

# EDA

In [ ]:
# Load the dataset and show basic info
df = pd.read_csv("annotations.csv")
# Display basic infor about columns
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49551 entries, 0 to 49550
Data columns (total 53 columns):
 #   Column                          Non-Null Count  Dtype 
---  ------                          --------------  ----- 
 0   Unnamed: 0                      49551 non-null  int64 
 1   person_id                       49551 non-null  int64 
 2   filename                        49551 non-null  object
 3   class1                          49551 non-null  object
 4   class2                          3677 non-null   object
 5   bounding_box                    49551 non-null  object
 6   gender_presentation_masc        49551 non-null  int64 
 7   gender_presentation_fem         49551 non-null  int64 
 8   gender_presentation_non_binary  49551 non-null  int64 
 9   gender_presentation_na          49551 non-null  int64 
 10  skin_tone_1                     49551 non-null  int64 
 11  skin_tone_2                     49551 non-null  int64 
 12  skin_tone_3                     49551 non-null

In [ ]:
# Show the data
df.describe()

Unnamed: 0     person_id  gender_presentation_masc  \
count  49551.000000  4.955100e+04              49551.000000   
mean   24775.000000  1.307137e+15                  0.670824   
std    14304.285931  1.293425e+15                  0.469919   
min        0.000000  3.501556e+14                  0.000000   
25%    12387.500000  6.396669e+14                  0.000000   
50%    24775.000000  8.475259e+14                  1.000000   
75%    37162.500000  1.314718e+15                  1.000000   
max    49550.000000  9.330696e+15                  1.000000   

       gender_presentation_fem  gender_presentation_non_binary  \
count             49551.000000                    49551.000000   
mean                  0.206757                        0.001917   
std                   0.404984                        0.043744   
min                   0.000000                        0.000000   
25%                   0.000000                        0.000000   
50%                   0.000000                        0.000000   
75%                   0.000000                        0.000000   
max                   1.000000                        1.000000   

       gender_presentation_na   skin_tone_1   skin_tone_2   skin_tone_3  \
count            49551.000000  49551.000000  49551.000000  49551.000000   
mean                 0.113055      0.130250      0.699824      0.940142   
std                  0.316664      0.443802      1.019083      1.101531   
min                  0.000000      0.000000      0.000000      0.000000   
25%                  0.000000      0.000000      0.000000      0.000000   
50%                  0.000000      0.000000      0.000000      1.000000   
75%                  0.000000      0.000000      1.000000      2.000000   
max                  1.000000      6.000000      6.000000      6.000000   

        skin_tone_4  ...  has_headscarf   has_eyeware  visible_torso  \
count  49551.000000  ...   49551.000000  49551.000000   49551.000000   
mean       0.879659  ...       0.046679      0.109624       0.728744   
std        1.030116  ...       0.210953      0.312424       0.444612   
min        0.000000  ...       0.000000      0.000000       0.000000   
25%        0.000000  ...       0.000000      0.000000       0.000000   
50%        1.000000  ...       0.000000      0.000000       1.000000   
75%        1.000000  ...       0.000000      0.000000       1.000000   
max        6.000000  ...       1.000000      1.000000       1.000000   

       visible_face  visible_minimal  lighting_underexposed  \
count  49551.000000     49551.000000           49551.000000   
mean       0.296099         0.148776               0.026498   
std        0.456540         0.355871               0.160612   
min        0.000000         0.000000               0.000000   
25%        0.000000         0.000000               0.000000   
50%        0.000000         0.000000               0.000000   
75%        1.000000         0.000000               0.000000   
max        1.000000         1.000000               1.000000   

       lighting_dimly_lit  lighting_well_lit   lighting_na  \
count        49551.000000       49551.000000  49551.000000   
mean             0.224092           0.797158      0.017719   
std              0.416987           0.402120      0.131930   
min              0.000000           0.000000      0.000000   
25%              0.000000           1.000000      0.000000   
50%              0.000000           1.000000      0.000000   
75%              0.000000           1.000000      0.000000   
max              1.000000           1.000000      1.000000   

       lighting_overexposed  
count          49551.000000  
mean               0.018991  
std                0.136493  
min                0.000000  
25%                0.000000  
50%                0.000000  
75%                0.000000  
max                1.000000  

[8 rows x 49 columns]

In [ ]:
# Check for missing values
df.isna().sum() # Only one column has missign entries and it's most cells

Unnamed: 0                            0
person_id                             0
filename                              0
class1                                0
class2                            45874
bounding_box                          0
gender_presentation_masc              0
gender_presentation_fem               0
gender_presentation_non_binary        0
gender_presentation_na                0
skin_tone_1                           0
skin_tone_2                           0
skin_tone_3                           0
skin_tone_4                           0
skin_tone_5                           0
skin_tone_6                           0
skin_tone_7                           0
skin_tone_8                           0
skin_tone_9                           0
skin_tone_10                          0
skin_tone_na                          0
age_presentation_young                0
age_presentation_older                0
age_presentation_middle               0
age_presentation_na                   0


In [ ]:
# Showing the column headers
df.columns

Index(['Unnamed: 0', 'person_id', 'filename', 'class1', 'class2',
       'bounding_box', 'gender_presentation_masc', 'gender_presentation_fem',
       'gender_presentation_non_binary', 'gender_presentation_na',
       'skin_tone_1', 'skin_tone_2', 'skin_tone_3', 'skin_tone_4',
       'skin_tone_5', 'skin_tone_6', 'skin_tone_7', 'skin_tone_8',
       'skin_tone_9', 'skin_tone_10', 'skin_tone_na', 'age_presentation_young',
       'age_presentation_older', 'age_presentation_middle',
       'age_presentation_na', 'hair_color_brown', 'hair_color_blonde',
       'hair_color_grey', 'hair_color_na', 'hair_color_black',
       'hair_color_colored', 'hair_color_red', 'hairtype_coily',
       'hairtype_dreadlocks', 'hairtype_bald', 'hairtype_straight',
       'hairtype_curly', 'hairtype_wavy', 'hairtype_na', 'has_facial_hair',
       'has_tattoo', 'has_cap', 'has_mask', 'has_headscarf', 'has_eyeware',
       'visible_torso', 'visible_face', 'visible_minimal',
       'lighting_underexposed', 'ligh

In [28]:
# Grouping gender together as an indexable list for altair
gender_cols = [
    "gender_presentation_masc", 
    "gender_presentation_fem", 
    "gender_presentation_non_binary", 
    "gender_presentation_na"
]
# Displaying the number of each type of presenting individuals
GenderCounts = df[gender_cols].sum().reset_index(name="count").rename(columns={"index": "gender"})
GenderCounts


gender  count
0        gender_presentation_masc  33240
1         gender_presentation_fem  10245
2  gender_presentation_non_binary     95
3          gender_presentation_na   5602

In [ ]:
# Using altair to graph the number of gender presentations (Sorted)
alt.Chart(GenderCounts).mark_bar().encode(
    x="count:Q",
    y= alt.Y("gender:N").sort('-x'),
    color="gender:N"
)


alt.Chart(...)

In [32]:
# Isolating the skintones
skin_cols = []
for column in df.columns:
    if column.startswith("skin_tone_"):
        skin_cols.append(column)

# Counting the number of different skin tones represented
skin_counts = df[skin_cols].sum().reset_index(name="count").rename(columns={"index": "skin_tone"})

# Displaying the number of different skin tones represented
alt.Chart(skin_counts).mark_bar().encode(
    x="count:Q",
    y= alt.Y("skin_tone:N").sort('-x'),
    color="skin_tone:N"
)


alt.Chart(...)

In [35]:
# Isolating the ages
age_cols = []
for column in df.columns:
    if column.startswith("age_presentation_"):
        age_cols.append(column)
        
# Counting the number of different ages represented
age_counts = df[age_cols].sum().reset_index(name="count").rename(columns={"index": "age_group"})

# Displaying the number of different ages represented
alt.Chart(age_counts).mark_bar().encode(
    x="count:Q",
    y= alt.Y("age_group:N").sort('-x'),
    color="age_group:N"
)


alt.Chart(...)

In [ ]:
# Isolating the hair color
hair_cols = []
for column in df.columns:
    if column.startswith("hair_color_"):
        hair_cols.append(column)
        
# Counting the number of different hairs colors represented
hair_counts = df[hair_cols].sum().reset_index(name="count").rename(columns={"index": "hair_color"})

# Displaying the number of different hairs colors represented
alt.Chart(hair_counts).mark_bar().encode(
    x="count:Q",
    y= alt.Y("hair_color:N").sort('-x'),
    color="hair_color:N"
)


alt.Chart(...)

In [42]:
hairtype_cols = [c for c in df.columns if c.startswith("hairtype_")]
hairtype_counts = df[hairtype_cols].sum().reset_index(name="count").rename(columns={"index": "hair_type"})

alt.Chart(hairtype_counts).mark_bar().encode(
    x="hair_type:N",
    y="count:Q",
    color="hair_type:N"
)

# Isolating the hair color
hair_type_cols = []
for column in df.columns:
    if column.startswith("hairtype_"):
        hair_type_cols.append(column)
        
# Counting the number of different hairs colors represented
hair_type_counts = df[hair_type_cols].sum().reset_index(name="count").rename(columns={"index": "hair_type"})

# Displaying the number of different hairs colors represented
alt.Chart(hair_type_counts).mark_bar().encode(
    x="count:Q",
    y= alt.Y("hair_type:N").sort('-x'),
    color="hair_type:N"
)



alt.Chart(...)

In [ ]:
# Gender masc + skin tone distribution
# Collecting all instances of masc presenting individuals
subset = df[df["gender_presentation_masc"] == 1]
# Counting the number of masc presenting individuals by skin tone
skin_counts_masc = subset[skin_cols].sum().reset_index(name="count").rename(columns={"index": "skin_tone"})

# Displaing number of different skin tones in the masc presenting group
alt.Chart(skin_counts_masc).mark_bar().encode(
    x="count:Q",
    y=alt.Y("skin_tone:N").sort('-x'),
    color="skin_tone:N"
).properties(title="Skin tone distribution (masculine-presenting)")


alt.Chart(...)

In [45]:
# Gender fem + skin tone distribution
# Collecting all instances of masc presenting individuals
subset = df[df["gender_presentation_fem"] == 1]
# Counting the number of masc presenting individuals by skin tone
skin_counts_fem = subset[skin_cols].sum().reset_index(name="count").rename(columns={"index": "skin_tone"})

# Displaing number of different skin tones in the masc presenting group
alt.Chart(skin_counts_fem).mark_bar().encode(
    x="count:Q",
    y=alt.Y("skin_tone:N").sort('-x'),
    color="skin_tone:N"
).properties(title="Skin tone distribution (feminine-presenting)")

alt.Chart(...)

In [20]:
heatmap_data = (
    skin_long.groupby(["class1", "skin_tone"]).size().reset_index(name="count")
)

alt.Chart(heatmap_data).mark_rect().encode(
    x="skin_tone:N",
    y="class1:N",
    color="count:Q"
).properties(title="Heatmap of class1 × Skin Tone")


alt.Chart(...)

# Dataset Documentation

### Dataset Overview

**Name:** Lucas Drager & Jack Manning 
**Owner/Contact:** [Class project team / instructor]  
**Access:** Class use only, internal distribution  

**Contents:**  
- **Items:** 49,551 annotated entries (person/image bounding boxes).  
- **Data fields:**  
  - Demographics: gender presentation (masc, fem, non-binary, NA), age presentation (young, middle, older, NA), skin tone (10-scale, NA).  
  - Appearance: hair color, hair type, presence of facial hair, tattoos, masks, headscarves, eyewear.  
  - Visual conditions: visibility of face/torso, lighting quality.  
  - Metadata: bounding box coordinates in JSON format, file name, class labels (e.g., “doctor,” “speaker”).  
- **Timeframe:** Static snapshot; timeframe not explicitly defined.  

**Intended Uses:**  
- Bias and fairness analysis in computer vision.  
- Exploratory data analysis (EDA).  

**Not Appropriate For:**  
- Production deployment in sensitive domains (healthcare, security).  
- Applications requiring balanced demographic representation without additional preprocessing.  

---

### Representativeness

- **Gender:** Strong skew toward masculine (33,240) and feminine (10,245) presentations, with very few non-binary (95).  
- **Skin tone:** Heavier representation of lighter tones (`skin_tone_1–3` total ~20k+) compared to darker tones (`skin_tone_8–10` < 500 combined).  
- **Age:** Most entries default to “not annotated,” limiting coverage.  
- **Other categories:** Some professions/roles are uneven (e.g., “speaker,” “doctor,” “lawman” dominate).  
- **Limitations:** Underrepresentation of marginalized groups makes the dataset less representative for fairness benchmarks; annotations are inferred rather than self-reported.  

---

### Data Quality

- **Missing data:** ~45,874 missing values, primarily in optional fields (hair type, lighting).  
- **Duplicates:** None detected.  
- **Noise:** Bounding boxes stored as JSON strings, not standardized tabular format.  
- **Potential Biases:**  
  - “NA” categories may bias models if treated as meaningful.  
  - Skew in demographics (esp. non-binary and darker skin tones) risks spurious correlations.  
- **Validation:** No clear evidence of external validation; labels appear to be manually or crowd-sourced, which can introduce subjectivity.  

---

### Pre-processing, Cleaning, Labeling

- Binary encodings used for most attributes (0/1).  
- Gender is effectively treated as categorical but functionally binary in practice.  
- Missing values retained as “NA” rather than imputed.  
- Bounding boxes not yet parsed into separate numerical fields.  

---

### Privacy

- Dataset includes **sensitive demographic information** (gender presentation, skin tone, age).  
- Risk of indirect re-identification if combined with other datasets.  
- No explicit consent or provenance documentation included (important limitation).  
